kong/cluster_events.lua

```diff
@@ -0,0 +1,334 @@
+  end
+
+  return true
+end
+
+
+local function poll(self)
+  -- get events since last poll
+
+  local min_at, err = self.shm:get(CURRENT_AT_KEY)
+  if err then
+    return nil, "failed to retrieve 'at' in shm: " .. err
+  end
+
+  if not min_at then
+    return nil, "no 'at' in shm"
+  end
+
+  -- apply grace period
+
+  min_at = min_at - self.poll_offset - 0.001
+
+  local max_at = ngx_now()
+
+  log(DEBUG, "polling events from: ", min_at, " to: ", max_at)
+
+  for rows, err, page in self.strategy:select_interval(self.channels, min_at, max_at) do
+    if err then
+      return nil, "failed to retrieve events from DB: " .. err
+    end
+
+    if page == 1 then
+      local ok, err = self.shm:safe_set(CURRENT_AT_KEY, max_at)
+      if not ok then
+        return nil, "failed to set 'at' in shm: " .. err
+      end
+    end
+
+    for i = 1, #rows do
+      local row = rows[i]
+
+      if row.node_id ~= self.node_id then
+        local ran, err = self.events_shm:get(row.id)
+        if err then
+          return nil, "failed to probe if event ran: " .. err
+        end
+
+        if not ran then
+          log(DEBUG, "new event (channel: '", row.channel, "') data: '",
+                     row.data, "' nbf: '", row.nbf or "none", "'")
+
+          local exptime = self.poll_interval + self.poll_offset
+
+          -- mark as ran before running in case of long-running callbacks
+          local ok, err = self.events_shm:set(row.id, true, exptime)
+          if not ok then
+            return nil, "failed to mark event as ran: " .. err
+          end
+
+          local cbs = self.callbacks[row.channel]
+          if cbs then
+            for j = 1, #cbs do
+              if not row.nbf then
+                -- unique callback run without delay
+                local ok, err = pcall(cbs[j], row.data)
+                if not ok and not ngx_debug then
+                  log(ERR, "callback threw an error: ", err)
+                end
+
+              else
+                -- unique callback run after some delay
+                local delay = max(row.nbf - ngx_now(), 0)
+
+                log(DEBUG, "delaying nbf event by ", delay, "s")
+
+                local ok, err = timer_at(delay, nbf_cb_handler, cbs[j], row)
+                if not ok then
+                  log(ERR, "failed to schedule nbf event timer: ", err)
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+  end
+
+  return true
+end
+
+
+if ngx_debug then
+  _M.poll = poll
+end
+
+
+local function get_lock(self)
+  local ok, err = self.shm:safe_add(POLL_LOCK_KEY, true, POLL_LOCK_EXPTIME)
+  if not ok and err ~= "exists" then
+    log(ERR, "failed to acquire poll lock: ", err)
+  end
+
+  return ok
+end
+
+
+poll_handler = function(premature, self)
+  if premature or not self.polling then
+    -- set self.polling to false to stop a polling loop
+    return
+  end
+
+  local ok, err = timer_at(self.poll_interval, poll_handler, self)
```

Comment:
a probably non-problem here might be that the `self` here will never be GC'ed. So if this `self` object might be renewed, and disposed of during the lifetime of the worker process, then there is a problem, otherwise not.